stdlib/public/Darwin/Accelerate/Quadrature.swift

```diff
@@ -0,0 +1,214 @@
+///     }
+///
+public class Quadrature {
+    
+    private var integrateOptions = quadrature_integrate_options()
+    private var integrand: ((Double) -> Double)!
+
+    /// Initializes and returns a quadrature instance.
+    ///
+    /// - Parameter integrator: An enumeration specifying the integration algorithm and relevant properties.
+    /// - Parameter absoluteTolerance: Requested absolute tolerance on the result.
+    /// - Parameter relativeTolerance: Requested relative tolerance on the result.
+    public init(integrator: Integrator,
+                absoluteTolerance: Double = 1.0e-8,
+                relativeTolerance: Double = 1.0e-2){
+
+        integrateOptions.abs_tolerance = absoluteTolerance
+        integrateOptions.rel_tolerance = relativeTolerance
+        
+        switch integrator {
+        case .qng:
+            integrateOptions.integrator = QUADRATURE_INTEGRATE_QNG
+        case .qag(let pointsPerInterval, let maxIntervals):
+            integrateOptions.integrator = QUADRATURE_INTEGRATE_QAG
+            integrateOptions.qag_points_per_interval = pointsPerInterval.rawValue
+            integrateOptions.max_intervals = maxIntervals
+        case .qags(let maxIntervals):
+            integrateOptions.integrator = QUADRATURE_INTEGRATE_QAGS
+            integrateOptions.max_intervals = maxIntervals
+        }
+    }
+    
+    /// The quadrature instance's requested absolute tolerance on the result.
+    public var absoluteTolerance: Double {
+        set {
+            integrateOptions.abs_tolerance = newValue
+        }
+        get {
+            return integrateOptions.abs_tolerance
+        }
+    }
+    
+    /// The quadrature instance's requested relative tolerance on the result.
+    public var relativeTolerance: Double {
+        set {
+            integrateOptions.rel_tolerance = newValue
+        }
+        get {
+            return integrateOptions.rel_tolerance
+        }
+    }
+ 
+    /// Performs the integration over the supplied function.
+    ///
+    /// - Parameter interval: The lower and upper bounds of the integration interval.
+    /// - Parameter integrand: The function to integrate. The input value is `x` that's within the interval over which the integrand is being integrated, and the output value is the corresponding value `y = integrand(x)` at those points.
+    public func integrate(over interval: ClosedRange<Double>,
+                          integrand: @escaping (Double) -> Double) -> Result {
+        
+        self.integrand = integrand
+        
+        let byteCount = MemoryLayout<Quadrature>.size
+        let integrandPointer = UnsafeMutableRawPointer.allocate(byteCount: byteCount,
+                                                                alignment: 1)
+        integrandPointer.storeBytes(of: self,
+                                    as: Quadrature.self)
+        
+        var callback = quadrature_integrate_function(
```

Comment:
If you passed the address of the integrand closure in directly, the closure would no longer need to escape, calling `integrate` wouldn't mutate any state, and `Quadrature` could become a thread-safe struct which just wraps the `quadrature_integrate_options`.

Reply:
I'm limited here: I need to pass the integrand via `fun_arg` (which is an `UnsafeMutableRawPointer`) because a C function pointer cannot be formed from a closure that captures context. However, I've significantly tidied up that code to avoid passing an instance of the class. (Pending push)

Reply:
That looks nice 👍 

It's true that you can't turn a capturing closure in to a C _function_ pointer, but you can turn it in to a void-pointer that you use again from Swift (which is what you're doing). It looks like you could now drop the `@escaping` (with help from `withoutActuallyEscaping` if the compiler moans at you) and make the whole thing a struct if you wanted to.

The nice Swift interface shouldn't really cost anything. It would be unfortunate if there were unnecessary heap allocations which made it more expensive than using C directly.

Reply:
Great points - I've changed `Quadrature` to a struct and used `withoutActuallyEscaping`. 

Many thanks!